datafusion/core/src/physical_optimizer/convert_first_last.rs

```diff
@@ -0,0 +1,294 @@
+                expr: groupby_exprs[idx].clone(),
+                options: None,
+            })
+            .collect::<Vec<_>>();
+
+        try_convert_first_last_if_better(
+            &requirement,
+            &mut aggr_expr,
+            input_eq_properties,
+        )?;
+
+        let required_input_ordering = (!requirement.is_empty()).then_some(requirement);
+
+        let input_order_mode =
+            if indices.len() == groupby_exprs.len() && !indices.is_empty() {
+                InputOrderMode::Sorted
+            } else if !indices.is_empty() {
+                InputOrderMode::PartiallySorted(indices)
+            } else {
+                InputOrderMode::Linear
+            };
+        let projection_mapping =
+            ProjectionMapping::try_new(group_by.expr(), &input.schema())?;
+
+        let cache = AggregateExec::compute_properties(
+            input,
+            plan.schema().clone(),
+            &projection_mapping,
+            mode,
+            &input_order_mode,
+        );
+
+        let aggr_exec = aggr_exec.new_with_aggr_expr_and_ordering_info(
+            required_input_ordering,
+            aggr_expr,
+            cache,
+            input_order_mode,
+        );
+
+        Ok(Transformed::yes(
+            Arc::new(aggr_exec) as Arc<dyn ExecutionPlan>
+        ))
+    } else {
+        Ok(Transformed::no(plan))
+    }
+}
+
+/// Get the common requirement that satisfies all the aggregate expressions.
+///
+/// # Parameters
+///
+/// - `aggr_exprs`: A slice of `Arc<dyn AggregateExpr>` containing all the
+///   aggregate expressions.
+/// - `group_by`: A reference to a `PhysicalGroupBy` instance representing the
+///   physical GROUP BY expression.
+/// - `eq_properties`: A reference to an `EquivalenceProperties` instance
+///   representing equivalence properties for ordering.
+/// - `agg_mode`: A reference to an `AggregateMode` instance representing the
+///   mode of aggregation.
+///
+/// # Returns
+///
+/// A `LexRequirement` instance, which is the requirement that satisfies all the
+/// aggregate requirements. Returns an error in case of conflicting requirements.
+///
+/// Similar to the one in datafusion/physical-plan/src/aggregates/mod.rs, but this
+/// function care only the possible conversion between FIRST_VALUE and LAST_VALUE
+fn try_convert_first_last_if_better(
+    prefix_requirement: &[PhysicalSortRequirement],
+    aggr_exprs: &mut [Arc<dyn AggregateExpr>],
+    eq_properties: &EquivalenceProperties,
+) -> Result<()> {
+    for aggr_expr in aggr_exprs.iter_mut() {
+        let aggr_req = aggr_expr.order_bys().unwrap_or(&[]);
+        let reverse_aggr_req = reverse_order_bys(aggr_req);
+        let aggr_req = PhysicalSortRequirement::from_sort_exprs(aggr_req);
+        let reverse_aggr_req =
+            PhysicalSortRequirement::from_sort_exprs(&reverse_aggr_req);
+
+        if let Some(first_value) = aggr_expr.as_any().downcast_ref::<FirstValue>() {
```

Comment:
Eventually (some other PR) it would be amazing if we can move this code into `FirstValue` somehow. As it is now, there is a coupling between the optimizer rule and the actual `PhysicalExpr` -- which means among other things this same optimization can't be used by user defined aggregates

Reply:
Yes, this should be similar to FunctionRewrite, registerable optimize rule.